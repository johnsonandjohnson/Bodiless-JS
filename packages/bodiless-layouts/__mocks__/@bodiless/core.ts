const editContext = {
  activate: jest.fn(),
  isEdit: true,
};

export const useEditContext = jest.fn(() => editContext);

const activateOnEffect = {
  setId: jest.fn(),
};
export const useActivateOnEffect = () => activateOnEffect;

export const useNode = jest.fn();

const contextMenuFormInner = jest.fn();
export const contextMenuForm = jest.fn(() => contextMenuFormInner);
