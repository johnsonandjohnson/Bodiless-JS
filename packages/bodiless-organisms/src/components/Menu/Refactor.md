# Navigation and Breadcrumbs refactor

## Goal
Improved DX and reduced LOE for activating navigation.

## High Level
- create new @bodiless/navigation package
  - possibly also create new @canvasx/navigation package, if necessary.
- deprecate SimpleMenu/MegaMenu.  Export a single menu which can be extended
  with different submenu types.
- clearly separate schema from styling
- include breadcrumb source logic by default in core menu hocs (it is needed
  for active trails even on sites without breadcrumbs).
- remove react-burger-menu library and replace with custom implementation.

> The specifications below are predicated on some lower level refactoring in
  in the [design api](https://github.com/wodenx/Bodiless-JS/blob/doc/post-pilot/packages/fclasses/doc/Refactor.md#tokens-httpsgithubcomjohnsonandjohnsonbodiless-jsissues826)
  
## Example API (brand level)

> In the docs below, tokens which are defined at brand level have a `$` prefix.
  Tokens defined at the site level have a $$ prefix.  All other methods are
  exported by @bodiless/navigation.

### Create the base schema

First we define the schema for menu titles. This is only required if you don't want to
use the default editors.

```js
export const $asEditableMenuTitle = asToken(
  asSchema,
  // Replace the link editor with a cusgtom one
  withDesign({
    Link: asToken(resetSchema, asMenuLink($asToggledLink)),
  }),
  // Extend the default text editor
  withDesign({
    Title: $asAutoSuperscript,
  }),
);
```

Note the use of `asMenuLink` above. It provides the node key structure. There is
an equivalent for the title (not used here bc we are just extending the default
editor).

Note also the use of `asSchema` and `resetSchema`.  These are "metatokens" -- objects
which attach or filter by token metadata.  This allows tokens which match certain
criteria to be selectively reset, so (as in this example) you can alter the
schema of a design key without altering its styling.  Both of these metatokens
would be defined at core library level, eg:


For megamenu, we can do the same for menu touts, if necessary...

```js
export const $asEditableMenuTout = asToken(
  asSchema,
  $withBrandToutEditors,
  $asEditableMenuTitle; // This replaces tout editors for title and link
);
```

The above would only be necessary if the brand had customized any of the default tout
editors.

Finally we define the menu schema itself:

```js
const $withBrandMenuSchema = asToken(
  asSchema,
  asBodilessMenu('MainMenu'),
  withListSubMenu($asEditableMenuTitle),
  withColumnSubMenu($asEditableMenuTitle),
  withToutSubMenu(withDesign($asEditableMenuTout),
  // It would be great to put this inside asBodilessMenu somehow, but we need
  // to know which submenus are applied 
  asBreadcrumbSource('List', 'Column', 'Tout'),
);
```

A few things to note here:
- `asBodilessMenu` is an extendsion `asBodilessList` which will turn a 'ul'
  into an editable menu.  It also defining the menu as a breadcrumb
  source (including wrapping top level menu items in `asBreadcrumb`).
- `with...SuBMenu` are helpers exported from @bodiless/navigation which add a
  submenu option to the menu. Each takes an optional token which will be applied
  to the sublist title. So, for example

  ```js
  withListSubMenu($asEditableMenuTitle)
  ```
  is equivalent to

  ```js
  asToken(
    asSchema,
    withListSubMenu(),
    withDesign({
      Item: withDesign({
        List: withDesign({
          Title: $asEditableMenuTitle,
        }),
      }),
    }),
  );
  ```
- `with...SubMenu` also takes care of wrapping each submenu item in `asBreadcrumb`.
- `withColumnSubmenu` manages the column sublists - so the title is applied to both.

  > Note: we want to make column submenus *optional* - so the items will have toggled sublists.


### Apply brand styling

```js
const $asBrandTopNav = asToken(
  asDesign, // A metatoken, like `asSchema`
  asTopNav('List', 'Columns'),
  $withBrandTopNavMainStyles,
  withSubMenuToken('List')($withBrandTopNavListStyles),
  withSubMenuToken('Columns')($withBrandTopNavColumnStyles),
  ...,
);
```

Note here:
- `asTopNav` is a refactor of the existing `asMegaMenuTopNav` to allow
  specifying which submenu types should have styling added. So
  ```js
  asTopNav('List')
  ```
  would be the same as our existing `asSimpleMenuTopNav`.
- `withSubMenuDesign` is a helper which makes it easier to target a particular
  type of submenu. The first parameter is a list of the submenu key(s) to which
  the token provided as a second argument should be applied. It also accepts the
  special key 'Main' to apply the design to the top level menu. So, for example,
  if you had common menu styles to apply to the top menu and all submenus, you
  could write:

  ```js
  withSubMenuToken('Main', 'List', 'Columns')($withCommonMenuStyles);
  ```


- Each of the `$withBrandTopNav...Styles` hocs should apply a simple list
  design, though the one for colukmn lists also has to design its sublist, eg.

```js
const $withTopNavColumnSubMenuStyles = asToken(
  asDesign,
  withDesign({
    Wrapper: withDesign({
      // We need to document clearly that the "Wrapper" of a sublist has 2 parts,
      // The <li> which is the outer wrapper and the <ul> which is the List.
      WrapperItem: $withSublistPadding,
      List: $withBullets,
    }),
    Item: ...,
    Title: ...,
  }),
  withSubMenuToken('List')(
    asDesign,
    withDesign({
      Wrapper: ...,
      Item: ...,
      Title: ...
    }),
  ),
);
```

### Same pattern for burger menu styling

```js
const $withBurgerMenuStyles = asToken(
  asDesign,
  // Same shorthand as `asTopNav` above.
  asBurgerMenu('List', 'Touts'),
  $withBurgerMenuMainStyles,
  withSubMenuToken('List')($withBurgerMenuListStyles),
  withSubMenuToken('Touts')($withBurgerMenuToutStyles),
  ...,
);
```

But, for Burger menus, we need to wrap the actual menu in the burger chrome. In order
to keep the api for burger and main menu as similar as possible, let's create an
HOC to do the wrapping, like

```js
const withBurgerMenuWrapper = (Component) => (props) => asToken(
  replaceWith(BurgerMenuClean),
  withDesign({
    // Note - in the current implementation this design key is 'Body' but we
    // should change it to 'Menu' (see note below).
    Menu: replaceWith(Component),
  })
);
```
Maybe we can even put this inside `asBurgerMenu` so we'd have:

```js
const $withBurgerMenuStyles = asToken(
  asDesign,
  $withBurgerMenuMainStyles,
  withSubMenuToken('List')($withBurgerMenuListStyles),
  ...,
);

cosnt $asBrandBurgerMenu = flow(
  asBurgerMenu('List', ...),
  // The design below applies to elements of the burger menu envelope.
  withDesign({
    Menu: $withBurgerMenuStyles,
    ...
  }),
);
```

> Note - in the current custom burger menu wrapper as implemented in the pilot
> we have 'Menu' and 'Body" design keys, which refer the the `<nav>` wrapper
> component and the actual menu respectively. We should rename these
to `Nav` (for the `<nav>` wrapper) and `Menu` (for the actual menu).

### Additional Burger Menu Considerations
- Refactor the way the overview link is added along
  [these lines](https://github.com/wodenx/Bodiless-JS/commit/26806e75bff973a2f8d231a458760cee1f33c676).
- Replace the existing burger menu implementation based on react-burger-menu
  with the one developed in scope of the pilots.
- We need a way to add additional menus/lists to the burger menu. Probably
  this can be done by embedding the main menu in another list, eg:

  ```js
  const asCustomBurgerMenu = (Menu) => (props) => {
    <ul>
      <li><Menu /></li>
      <li>{...extraMenus}</li>
    </ul>
  };
  ```

  Alternatively we could inject our custom menus into the main menu using the
  [same pattern as the overview link]((https://github.com/wodenx/Bodiless-JS/commit/26806e75bff973a2f8d231a458760cee1f33c676).

  ```js
  const asCustomBurgerMenu = (Menu) => (props) => {
    <Menu>
      {...extraMenus}
    </Menu>
  }
  ```

### Responsive variants

After creating `$asBrandBurgerMenu` and `$asBrandTopNav` above, we're left
with the following, which is probably fine.

```js
cosnt $withBrandMenuVariants = asToken(
  withResponsiveVariants({
    _default: $asBrandBurgerMenu,
    lg: $asBrandTopNav,
  }),
  // Note - the css visibility applied below could be bundled into
  // the $asSiteBurgerMenu and $asSiteTopNav,
  withDesign({
    _default: $asMobileOnly,
    lg: $asDesktopOnly,
  }),
);

const $asBrandNavigation = asToken(
  $withBrandMenuSchema,
  $withBrandMenuVariants,
);
```

THis token can then be applied directly to the 'Navigation' key in the layout (the
default component for that key would probalby be `Ul`):
```js
const $asSiteLayout = withDesign({
  Navigation: $asBrandNavigation,
});
```

We can also create a helper at brand or global level to simplify applying the same
token to both menus:
```js
const $withMenuToken = (...tokenDef: TokenDef[]) = withDesign({
  _default: asToken(...tokenDef),
  lg: withDesign({
    Menu: asToken(...tokenDef),
  })
});
```


### Customizing

Some examples of how local customizations could be achieved. Tokens defined locally
are prefixed with double `$$`.

#### Swapping out a title editor

```js
// First create a token which applies to the main menu and all submenus. 
const $$withSiteMenuEditor = withSubMenuToken('Main', 'List', 'Columns')(
  withDesign({
    Title: withDesign({
      Title: asToken(resetSchema, $$withCustomEditor)
    }),
  }),
};

const $$asSiteLayout = asToken(
  $asBrandLayout,
  withDesign({
    Navigation: $withMenuToken($$withSiteMenuEditor)
  });
);
```

### Changing the style of a menu tout

```js
// Define our custom tout styling
const $$withSiteMenuToutStyles = flow(
  asToutVertical,
  $withToutDefaultStyle,
  ...,
);

// create a submenu design which replaces default tout with custom one.
const $$withSiteCustomMenuTout = withSubMenToken('Touts')(
  withDesign({
    Title: asToken(
      $resetAllButSchema, // or resetDesign if we have more than 2 domains?
      $$withSiteMenuToutStyles,
    ),
  }),
);

// Inject this into the layout
const $$asSiteLayout = withDesign({
  Navigation: withDesign({
    // We only want to apply this to desktop...
    lg: $$withSiteCustomMenuTout,
  }),
});
```

## Breadcrumbs

The primary goals here are:
- Clearly separate schema from styling
- Provide defaults so that the most common use-cases require minimal configuration
- Populate the breadcrumb store by default for all menus.

### Breadcrumb source and store.

The current `asBreadcrumbSource` hoc  should be decomposed and bundled into the menu hocs,
as described above.

The [Breadcrumb Store Provider](https://github.com/johnsonandjohnson/Bodiless-JS/blob/master/examples/test-site/src/components/Layout/index.tsx#L46)
should be dealt with in scope of creating a designable Layout component

### Breadcrumb component 

Here, the primary consideration is to separate schema from styling, and to provide
defaults so that the most common use-cases require minimal configuration. 
which are currently site-level to the package.

> Note: we need to preseve all existing design keys in the current breadcrumb
> component

- Remove
  [the pre-connected `Breadcrumbs` component](https://github.com/johnsonandjohnson/Bodiless-JS/blob/master/packages/bodiless-components/src/Breadcrumbs/Breadcrumbs.tsx#L312-L315)
  and instead export an `asMenuBreadcrumbs` which is basically just
  ```js
  const asMenuBreadcrumbs = flow(
    withBreadcrumbItemsFromStore,
    observer,
  );
  ```
- Package withEditableStartingTrail, withEditableFinalTrail and
  withBreadcrubmSchema with default editors similar to what we do with the menu
  items.
- Make it possible to apply the same `$asEditableMenuTitle` token to breadcrumbs
  as is applied for menu titles, so that if a site builder needs to change the
  schema, she can do it in one place.
- Make it possible to style all breadcrumb items with a single token, something
  like
  ```
  const withBreadcrumbItemToken = (...tokenDefs: TokenDef[]) => withDesign({
    StartingTrail: asToken(...tokenDefs),
    EndingTrail: asToken(...tokenDefs),
    BreadcrumbLink: asToken(...tokenDefs),
  });
  ```
  All three of these should have access to a node with the same schema, so that
  active trail effects can be applied, or other styling choices made based on
  the link data.

## Suggested Bodiless Refactors/New Functions

### Metatokens

These metatokens should probably be defined at a bodiless core level (or in a
core design system library).

```js
export const asSchema: MetaToken<any> = {
  meta: {
    categories: {
      Domain: 'Schema',
    }
  }
};
export const resetSchema: MetaToken<any> = {
  ...asSchema,
  filter: (h: Token<any>) => !h.meta?.categories?.Domain?.includes('Schema'),
}
```

### Menu Titles

```js
const MenuTitleBase = ({ components: C, ...rest }) => (
  <C.Link {...rest}><C.Title /><C.Link>
);
export const MenuTitle = designable({ Link: A, Title: Fragment }, 'MenuTitle')(MenuTitleBase);

export const asMenuLInk = (asEditableLink) => asToken(
  asSchema,
  withSidecarNodes(
    asEditableLink('link', undefined, () => ({ groupLabel: 'Menu Item' })
  ),
);
export const asMenuTitle = (asEditable) => asToken(
  asSchema,
  asEditable('title', 'Menu Item'),
);
export const asEditableMenuTitle = asToken(
  asSchema,
  withDesign({
    Link: asMenuLink(asBodilessLinkToggle(asBodilessLink, Div))()),
    Title: asMenuTitle(asEditable(undefined, 'Menu Item')),
  }),
  withNode,
  withNodeKey('title'),
);
```

## asBodilessMenu

```js
export const asBodilessMenu = (nodeKeys?: WithNodeKeyProps, defaultData, useOverrides) => asToken(
  asSchema,
  asBodilessList(
    nodeKeys,
    defaultData,
    (props) => ({ groupLabel: 'Menu Item', ...useOverrides(props) })
  ),
  asStylableList,
  withDesign({
    Item: asChameleonSubList(() => ({ formTitle: 'Sub-Menu Type' })),
  }),
  withMenuContext,
  withDesign({
    Title: startWith(MenuTitleClean);
  }),
  // Maybe abstract the following into `withEditableMenuTitle` which takes nodeKeys?
  withDesign({
    Title: asEditableMenuTitle,
    Item: asBreadcrumb({
      // Note we hardcode these node keys. They are also hardcoded in asEditableMenuTitle.
      linkNodeKey: 'title$link',
      titleNodeKey: 'title$title',
    }),
  }),
);
```

### Design helpers

```js
const withSubMenuToken = (...keys: string[]) => (...tokenDefs: TokenDef[]) => {
  const design = keys.reduce((acc, next) => ({
    ...acc,
    [key]: asToken(tokenDefs),
  }));
  const withSubMenuDesign = withDesign({
    Item: withDesign(omit(design, 'Main')),
  });
  if (design['Main']) return asToken(
    design['Main'],
    withSubMenuDesign,
  );
  return withSubMenuDesign;
}
```

## Scratchpad

```js
const withMenuTitleEditors = design => {
  

}
const asMenuTout = ({ linkNodeKey = 'link', textNodeKey = 'text' }) => {
  // We apply the node keys ourselves
  const transformDesign = design => {
    const Link = flow(
      replaceable,
      withSidecarNodes(
        design.Link || withBodilessLinkToggle(asBodilessLink, Div)(),
        withNodeKey(linkNodeKey),
      ),
    );
    const Title = flow(
      replaceable,
      design.Text || asEditable(undefined, 'Menu Item'),
      withNodeKey(textNodeKey),
    );
    return { ...design, Link, Title };
  };
  const ToutClean
  return startWith(flow(
    replaceable
    startWith(({ design, ...rest}) => <ToutClean design={transformDesign(design)} {...rest} />);
    withNode,
    withNodekey('title'),
  )(ToutClean));
}

const MenuLinkBase = ({ components:C, ...rest }) => <C.Link {...rest}><C.Text /></C.Link>;
const MenuLink = ({ linkNodeKey = 'link', textNodeKey = 'text' }) => {
  const applyDesign = design => {
    const Link = flow(
      replaceable,
      withSidecarNodes(
        design.Link || withBodilessLinkToggle(asBodilessLink, Div)(),
        withNodeKey(linkNodeKey),
      ),
    )(A);
    const Title = flow(
      replaceable,
      design.Title || asEditable(undefined, 'Menu Item'),
      withNodeKey(textNodeKey),
    )(Fragment);
    return { Link, Text };
  };
  return startWith(flow(
    designable(applyDesign, 'MenuLink'),
    stylable,
    withNode,
    withNodeKey('title'),
  )(MenuLinkBase));
};
const withSubMenuDesign = design => withDesign({
  Item: withDesign(design),
});

const withMenuDesign = keys => design => flow(
  withDesign(design),
  ...keys.map(key => withSubMenuDesign({
    [key]: design,
  })),
);

const asMenuSubList = flow(asMenuSublistOld, withDesign({ Item: asBreadcrumb }));

const withListSubMenu = withSubMenuDesign({
  List: asMenuSubList
});

const withToutSubMenu = withSubMenuDesign({
  Touts: asMenuSubList,
});

const withColumnSubMenu = withSubMenuDesign({
  Columns: flow(
    asMenuSubList,
    withDesign({
      Item: asMenuSubList,
    }),
  ),
});

// In fclasses?
type SlotComponents = {
  Component: ComponentType<any>,
};
type SlotProps = {
  components: SlotComponents,
};
const Slot = ({ components: C, ...rest}: SlotProps) => <C.Component {...rest} />;

type MenuLinkComponents = {
  Link: CT<any>,
  Text: CT<any>,
};

const $withItemSchema = withDesign({
  Title: flow(
    asMenuLink,
    withDesign({})
  )
})


// This is inside asMenu, withLinkSubMenu, etc...
const withDefaultItem = withDesign({
  Title: flow(
    asMenuLink,
    withDesign({
      LinkEditor: asBodilessLinkToggle(asBodilessLink)(),
      TextEditor: asEditable(undefined, 'Menu Item'),
    }),
  ),
  Item: asBreadcrumb({ linkNodeKey: 'title$link', titleNodeKey: 'title$text' })
});

const $withBrandMenuEditors = withDesign({
  Title: withDesign({
    // Here we extend the default editor
    TextEditor: asAutoSuperscript,
    // Here we replace the default editor
    // Note - we don't provide a nodeKey - that will be provided
    // by the default 
    LinkEditor: flow(reset, asCustomEditableLink())
  })
});

const $withMenuSchema = flow(
  // Provide the nodeKey for the whole menu.
  asMenu('MainMenu'),
  // Add menu editors if necessary.
  $withBrandMenuEditors,
  // This runs 
  withListSubMenu,
  withToutSubMenu,
  withSubMenuDesign({
    List
  })
  $withExtendedTitleEditor,
  $withReplaceLinkEditor
)

flow(
  asMenu(),
  withDefaultItem,

  // Or
  withDesign({
    Title: withDesign({
      Link: reset(asSpecialEditable()),
      Text: reset(asSpecialEditableLink()),
    }),
  })
)

const asMenuTout = (withToutEditors, nodeKeys)

const asMenuLink = (nodeKeys = _.keyBy('link', 'text')) => {
  const applyDesign = design => {
    const Component flow(
      withSidecarNodes(flow(
        design.Link,
        withNodeKey(nodeKeys.link),
      ),
      stylable,
      design.Text,
      withNodeKey(nodeKeys.text),
      withNode,
      withNodeKey('title'),
    )(A),
    return { Component };
  };
  return designable<MenuLinkComponents>(applyDesign)(Slot);
}
```
