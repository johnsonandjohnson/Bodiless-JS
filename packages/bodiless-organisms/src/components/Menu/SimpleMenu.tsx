/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { flow } from 'lodash';

import {
  withDesign,
  Design,
  stylable,
  HOC,
} from '@bodiless/fclasses';
import {
  WithNodeKeyProps,
} from '@bodiless/core';
import {
  asBodilessList,
  withSubListDesign, withSubLists, asSubList, withDeleteNodeOnUnwrap,
  asBreadcrumbSource as asBreadcrumbSourceBase,
} from '@bodiless/components';

import { asStylableList } from './SimpleMenu.token';
import withMenuContext from './withMenuContext';

/**
 * Creates a stylable sublist which deletes it's data when the last item is removed.
 * Suitable for use for all menus.
 */
const asMenuSubList = flow(
  asSubList(() => ({ groupLabel: 'Sub-Menu Item' })),
  asStylableList,
  withDesign({
    Wrapper: withDesign({
      WrapperItem: stylable,
      List: stylable,
    }),
  }),
  withDeleteNodeOnUnwrap('sublist'),
);

/**
 * Applies the specified design to the main list and all sublists.
 *
 * @param design
 */
const withMenuDesign = (design: Design<any>|HOC) => {
  const withDesign$ = typeof design === 'function' ? design : withDesign(design);
  return flow(
    withSubListDesign(1)({
      SubMenu: withDesign$,
    }),
    withDesign$,
  );
};

/**
 * Bodiless HOC generator which creates the basic structure of the Mega Menu. The component
 * to which the HOC applies is irrelevant (it will be replaced by the Menu wrapper).
 *
 * The base mega menu serves as a base for various views on the Menu data, including
 * a site's main menu, a burger menu and breadcrumbs.
 *
 * @param nodeKeys The optional nodekeys specifying where the data should be stored.
 *
 * @return HOC which creates a basic mega menu list.
 */
const asMenuBase = (nodeKeys?: WithNodeKeyProps) => flow(
  asBodilessList(nodeKeys, undefined, () => ({ groupLabel: 'Menu Item' })),
  asStylableList,
  withSubLists(1)({ SubMenu: asMenuSubList }),
  withMenuContext,
);

const asBreadcrumbSource = asBreadcrumbSourceBase(withMenuDesign);

export {
  asMenuBase, asBreadcrumbSource, withMenuDesign, asMenuSubList,
};
