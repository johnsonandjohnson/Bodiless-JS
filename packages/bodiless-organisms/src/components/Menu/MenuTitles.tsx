/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { flow, identity } from 'lodash';
import {
  withNode, withNodeKey, withSidecarNodes,
} from '@bodiless/core';
import { replaceWith, HOC, stylable } from '@bodiless/fclasses';
import {
  asBodilessLink, withBodilessLinkToggle,
} from '@bodiless/components';
import { ToutClean } from '../Touts';
import MenuLink from './MenuLink';

export const asMenuTout = (withToutEditors: any) => flow(
  replaceWith(ToutClean),
  withToutEditors,
  withNode,
  withNodeKey('title'),
);

export const asMenuLink = (asEditable: HOC) => flow(
  replaceWith(MenuLink),
  withSidecarNodes(
    withBodilessLinkToggle(asBodilessLink)('link'),
  ),
  stylable,
  asEditable,
  withNode,
  withNodeKey('title'),
) as HOC;

export const asDefaultMenuLink = asMenuLink(identity);

export const asDefaultMenuTout = asMenuTout(identity);
