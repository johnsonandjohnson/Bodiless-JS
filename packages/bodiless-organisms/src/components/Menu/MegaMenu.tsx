/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { flow } from 'lodash';

import {
  withDesign, HOC,
} from '@bodiless/fclasses';
import { WithNodeKeyProps } from '@bodiless/core';
import {
  asBodilessList, asChameleonSubList, asBreadcrumbSource as asBreadcrumbSourceBase,
} from '@bodiless/components';

import { asMenuSubList } from './SimpleMenu';
import { asStylableList } from './SimpleMenu.token';
import withMenuContext from './withMenuContext';

/**
 * Applies a list design (or other HOC) recursively to all submenus.
 *
 * @param design The design object or HOC to be applied.
*/
const withSubMenuDesign = (design: any) => {
  const withDesign$ = typeof design === 'function' ? design : withDesign(design);
  return withDesign({
    Item: withDesign({
      List: withDesign$,
      Touts: withDesign$,
      Columns: flow(
        withDesign$,
        withDesign({
          Item: withDesign$,
        }),
      ),
    }),
  });
};

/**
 * Applies a list design (or other HOC) to the main menu and all submenus.
 *
 * @param design The design object or HOC to be applied.
*/
const withMenuDesign = (design: any): HOC => {
  const withDesign$ = typeof design === 'function' ? design : withDesign(design);
  return flow(
    withSubMenuDesign(withDesign$) as HOC,
    withDesign$ as HOC,
  );
};

/**
 * Bodiless HOC generator which creates the basic structure of the Mega Menu. The component
 * to which the HOC applies is irrelevant (it will be replaced by the Menu wrapper).
 *
 * The base mega menu serves as a base for various views on the Menu data, including
 * a site's main menu, a burger menu and breadcrumbs.
 *
 * @param nodeKeys The optional nodekeys specifying where the data should be stored.

 * @return HOC which creates a basic mega menu list.
 */
const asMenuBase = (nodeKeys?: WithNodeKeyProps) => flow(
  asBodilessList(nodeKeys, undefined, () => ({ groupLabel: 'Menu Item' })),
  asStylableList,
  withDesign({
    Item: asChameleonSubList(() => ({ formTitle: 'Sub-Menu Type' })),
  }),
  withSubMenuDesign(asMenuSubList),
  withMenuContext,
);

const asBreadcrumbSource = asBreadcrumbSourceBase(withMenuDesign);

/**
 * Creates a HOC which can be applied to a base menu to make it into a data source for
 * the site's breadcrumbs.
 *
 * @param settings The title and link nodekeys defining where to locate the link and title nodes.
 *
 * @return  HOC for providing breadcrumb data from this menu.
 */
export {
  asMenuSubList, asMenuBase, withMenuDesign, asBreadcrumbSource,
};
