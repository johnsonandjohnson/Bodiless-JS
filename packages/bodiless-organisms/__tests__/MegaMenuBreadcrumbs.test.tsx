/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { mount, ReactWrapper } from 'enzyme';
// eslint-disable-next-line import/no-extraneous-dependencies
import cheerio from 'cheerio';
import { withDefaultContent, withSidecarNodes } from '@bodiless/core';
import {
  asBodilessLink, asEditable, withBreadcrumbStore, Breadcrumbs,
} from '@bodiless/components';
import { replaceWith, withDesign } from '@bodiless/fclasses';
import { flowRight } from 'lodash';

import { asMenuBase, asBreadcrumbSource } from '../src/components/Menu/MegaMenu';

const { DefaultContentNode } = require('@bodiless/core');

const setPagePath = (pagePath: string) => {
  Object.defineProperty(DefaultContentNode.prototype, 'pagePath', {
    value: pagePath,
    writable: false,
  });
};
const createBreadcrumbComponent = ({
  content = {},
}) => {
  const Source = flowRight(
    asBreadcrumbSource({
      linkNodeKey: 'title$link',
      titleNodeKey: 'title$text',
    }),
    asMenuBase('testMenu'),
  )('ul');

  const BreadcrumbComponent = (props: any) => (
    <>
      <Source />
      <Breadcrumbs {...props} />
    </>
  );

  return flowRight(
    withDefaultContent(content),
    withDesign({
      // @ts-ignore
      BreadcrumbLink: replaceWith(withSidecarNodes(
        asBodilessLink(),
      ))('a'),
      BreadcrumbTitle: replaceWith(
        asEditable()(React.Fragment),
      ),
    }),
    withBreadcrumbStore,
  )(BreadcrumbComponent);
};

// Helper to get the html of only the breadcrumbs.
const breadcrumbHtml = (wrapper: ReactWrapper) => {
  // We use cheerio directly here bc using enzyme's 'render'
  // method generates warnings about not using layout
  // effects on the server.
  const $ = cheerio.load(wrapper.html());
  return $.html($('body>ul').last());
};

const generateMegaMenuContent = (component: string) => ({
  testMenu: {
    items: [
      'home',
      'products',
    ],
  },
  testMenu$home$title$link: {
    href: '/',
  },
  testMenu$home$title$text: {
    text: 'home',
  },
  testMenu$products$title$link: {
    href: '/products',
  },
  testMenu$products$title$text: {
    text: 'products',
  },
  testMenu$products$sublist: {
    items: [
      'productA',
      'productB',
    ],
  },
  'testMenu$products$cham-sublist': {
    component,
  },
  testMenu$products$sublist$productA$title$link: {
    href: '/products/productA',
  },
  testMenu$products$sublist$productA$title$text: {
    text: 'ProductA',
  },
  testMenu$products$sublist$productB$title$link: {
    href: '/products/productB',
  },
  testMenu$products$sublist$productB$title$text: {
    text: 'ProductB',
  },
});

describe('asBreadcrumbsClean', () => {
  it('creates breadcrumbs for 2-level Touts MegaMenu', () => {
    setPagePath('/products/productA');
    const Breadcrumb = createBreadcrumbComponent({
      content: generateMegaMenuContent('Touts'),
    });
    const wrapper = mount(<Breadcrumb />);
    expect(breadcrumbHtml(wrapper)).toMatchSnapshot();
  });
  it('creates breadcrumbs for 2-level List MegaMenu', () => {
    setPagePath('/products/productA');
    const Breadcrumb = createBreadcrumbComponent({
      content: generateMegaMenuContent('List'),
    });
    const wrapper = mount(<Breadcrumb />);
    expect(breadcrumbHtml(wrapper)).toMatchSnapshot();
  });
  it('creates breadcrumbs for 2-level Columns MegaMenu', () => {
    setPagePath('/products/productA');
    const Breadcrumb = createBreadcrumbComponent({
      content: generateMegaMenuContent('Columns'),
    });
    const wrapper = mount(<Breadcrumb />);
    expect(breadcrumbHtml(wrapper)).toMatchSnapshot();
  });
  it('creates breadcrumbs for 3-level Columns MegaMenu', () => {
    setPagePath('/products/productA/subProduct');
    const Breadcrumb = createBreadcrumbComponent({
      content: {
        ...generateMegaMenuContent('Columns'),
        testMenu$products$sublist$productA$sublist: {
          items: [
            'subProduct',
          ],
        },
        testMenu$products$sublist$productA$sublist$subProduct$title$link: {
          href: '/products/productA/subProduct',
        },
        testMenu$products$sublist$productA$sublist$subProduct$title$text: {
          text: 'SubProduct',
        },
      },
    });
    const wrapper = mount(<Breadcrumb />);
    expect(breadcrumbHtml(wrapper)).toMatchSnapshot();
  });
});
