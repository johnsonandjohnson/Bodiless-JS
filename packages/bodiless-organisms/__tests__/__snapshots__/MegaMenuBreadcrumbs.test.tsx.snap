// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asBreadcrumbsClean creates breadcrumbs for 2-level Columns MegaMenu 1`] = `"<ul><li><a href=\\"/products/\\"><span>products</span></a></li><span></span><li><span>ProductA</span></li></ul>"`;

exports[`asBreadcrumbsClean creates breadcrumbs for 2-level List MegaMenu 1`] = `"<ul><li><a href=\\"/products/\\"><span>products</span></a></li><span></span><li><span>ProductA</span></li></ul>"`;

exports[`asBreadcrumbsClean creates breadcrumbs for 2-level Touts MegaMenu 1`] = `"<ul><li><a href=\\"/products/\\"><span>products</span></a></li><span></span><li><span>ProductA</span></li></ul>"`;

exports[`asBreadcrumbsClean creates breadcrumbs for 3-level Columns MegaMenu 1`] = `"<ul><li><a href=\\"/products/\\"><span>products</span></a></li><span></span><li><a href=\\"/products/productA/\\"><span>ProductA</span></a></li><span></span><li><span>SubProduct</span></li></ul>"`;
